### PR TITLE
feat(post): kubernetes gated deployments

### PR DESCRIPTION
**Pull request by stevenkfu**

### CODE REVIEW

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    url:
+    photo:
+  - name: Silas Boyd-Wickizer
+    title: Sr. Director of Engineering
+    url: https://github.com/silasbw
+    photo: https://avatars.githubusercontent.com/silasbw
+---
+
+At GoDaddy, teams frequently perform continuous deployments of their services to
+Kubernetes. When engineers make changes to these services, those changes may
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: Changes to a service would always change the behavior of that service, right? If your change had no effect on the behavior of the service, why would you make the change?

I think what you mean to say is that "Changes to a service may introduce unintended changes that weren't caught by unit and integration tests, resulting in a negative..."

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+affect the behavior of the application, resulting in a negative impact on the
+customer experience that unit and integration tests would not catch. For
+example, an engineer might introduce a performance regression that increases
+page load time. In this case, it is important to mitigate the impact on
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: Can you provide a citation for "it is important"? It's generally understood that performance ties to revenue but I think it delivers a stronger message to quantify that (or at least show the research that backs your opinion)

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+customers by promptly removing the change from production. Measuring the impact
+of a change and making a quick decision to roll back can be challenging in an
+environment where there are many deploys each day. To help address this
+challenge, we designed, implemented, and open sourced an extension for
+Kubernetes, called Kubernetes Gated Deployments, that automates regression
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: This describes the outcome, but doesn't go into detail on how you arrived here. Did you evaluate industry alternatives? What metrics did you care about when making those evaluations, and how did those different tools measure up? 

tl;dr Why should someone consider your solution vs other well-known continuous delivery tools like Spinnaker?

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+undesirable behavior.
+
+## Challenges in analyzing application metrics
+
+When changes impact application metrics, engineers need to consider if rolling
+back is necessary. However, it is difficult to make data-driven decisions that
+are statistically sound, and that process is often time-consuming, leading to
+prolonged customer impact. Common techniques for assessing the impact of changes
+on key metrics of an application's functionality and performance include
+pre-post analysis, load testing and benchmarking, and canary deploys with manual
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: Where would I go to learn more about these techniques?

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+inspection of metrics. There are several drawbacks with these approaches:
+
+* They are not scalable, and do not isolate and measure the impact of the change
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: nit: Is "scalable" really the problem here, or is it some other factor? If I hired a release team and had them focus on this specific problem, I could scale that team to solve this particular problem, right?

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+  traffic pattern similar enough to real production traffic to guarantee the
+  load test captures what engineers intend to measure.
+* It is difficult to monitor and respond to regressions in key metrics in a
+  timely manner: Teams may rely on monitors to show increased response times or
+  complaints from users, but at this point the change has already adversely
+  affected the customer experience, and engineers must roll back the change.
+
+## Kubernetes Gated Deployments
+
+To solve the problems described above, we’ve created Kubernetes Gated
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: nit: "we've" is using smart quotes (`’`) instead of apostrophes (`'`)

Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: nit: Run-on sentence.

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+either rolls back the change or promotes it to take all production traffic.
+Kubernetes Gated Deployments does not require any application code changes, and
+relies on native Kubernetes resources to route traffic and manage the deployed
+application. It adds a new object type, called a `GatedDeployment`, that allows
+engineers to specify metric sources and analysis techniques, which the
+controller will use to compare versions. We designed this system to augment
+existing deployment processes by comparing the deployed versions to validate
+that the new one does no harm to the metrics specified in that object.
+
+## How it works
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: I think there's an opportunity here to dig into the details of the developer experience here - what happens when there are multiple changes, or changes that depend on each other? What happens if I don't have a lot of traffic?

eg: 20 commits happen at about the same time; how long does it take for those changes to reliably get to prod through GatedDeployments, what happens if something goes wrong, and what's the likelihood that it identifies the commit with the defect?

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 8]
```diff
@@ -0,0 +1,245 @@
+---
+layout: post
+title: "Kubernetes Gated Deployments"
+date: 2019-08-08 09:00:00 -0700
+cover: /assets/images/kubernetes-gated-deployments/cover.jpg
+excerpt: Kubernetes Gated Deployments is a Kubernetes controller that facilitates automatic regression testing and canary analysis on Kubernetes deployments, that can detect and roll back the causal change without developer intervention. It is designed to augment existing deployment processes by analyzing key functionality and performance metrics associated with the application.
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: nit: Lots of commas in the first line of that excerpt.

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+exposes a threshold setting before it marks the treatment version as causing
+harm to the metric of response time. This is useful in the case when the
+application does not receive much traffic, as small sample sizes can result in
+erroneous statistically significant results. Depending on the traffic patterns
+and response time distribution of the application, teams can also configure a
+minimum number of samples before analyzing the response times, and a maximum
+amount of time in seconds that the experiment will run for before automatically
+deciding that it is a success.
+
+## Kubernetes Gated Deployments in production
```
Comment (gisenberg) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: It'd be nice to see a visualization of what pass/fail looks like here - what does New Relic show for a positive case and a negative case?

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+    title:
+    url:
+    photo:
+  - name: Silas Boyd-Wickizer
+    title: Sr. Director of Engineering
+    url: https://github.com/silasbw
+    photo: https://avatars.githubusercontent.com/silasbw
+---
+
+At GoDaddy, teams frequently perform continuous deployments of their services to
```
[the remark this answers is not in the document]
Comment (jeffpearce) [reply _posts/2019-08-08-kubernetes-gated-deployments.md]: 👋

File: _posts/2019-08-08-kubernetes-gated-deployments.md [hunk 11]
```diff
@@ -0,0 +1,245 @@
...
+negative impact. Kubernetes Gated Deployments does not require any application
+code changes, making it simple to apply to existing deployment pipelines.
+
+If you want to learn more about or get started with Kubernetes Gated
+Deployments, please visit
+<https://github.com/godaddy/kubernetes-gated-deployments>. We welcome
+contributions and would love to expand the decision plugins available to support
+other metric backends and analysis techniques.
+
+The GoDaddy Machine Learning team built Kubernetes Gated Deployments. Interested
```
Comment (silasbw) [inline _posts/2019-08-08-kubernetes-gated-deployments.md]: GoDaddy Machine Learning and Experimentation team built ... ?

File: _posts/2019-08-13-kubernetes-gated-deployments.md [hunk 12]
```diff
@@ -0,0 +1,280 @@
...
+Deployments, please visit
+<https://github.com/godaddy/kubernetes-gated-deployments>. We welcome
+contributions and would love to expand the decision plugins available to support
+other metric backends and analysis techniques.
+
+Our team of engineers and full-stack data scientists builds experimentally
+validated services. Come join us and contribute to services that help us
+leverage experimentation services and machine learning models to improve the
+GoDaddy customer experience. We're looking for a
+[principal](https://godaddy.wd1.myworkdayjobs.com/GoDaddy_careers/job/Kirkland/Principal-Software-Engineering----Machine-Learning-and-Experimentation_R006162-1)
```
Comment (ealun) [inline _posts/2019-08-13-kubernetes-gated-deployments.md]: @stevenkfu Sorry, I gave you the wrong links. Can you swap in these instead please?

https://careers.godaddy.com/job/cambridge/principal-software-development-engineer-machine-learning-and-experimentation/18045/11812427

https://careers.godaddy.com/job/cambridge/senior-software-engineering-machine-learning-and-experimentation/18045/12444210 and

Comment (stevenkfu) [reply _posts/2019-08-13-kubernetes-gated-deployments.md]: updated
